<?xml version="1.0" ?>
<launch>
  
  <arg name="bag" default="false"/>
  <arg name="bag_file" default=""/>
  <arg name="bag_start" default="0"/>
  <arg name="camera" default="front"/>

  <!-- Load bag file data -->
  <group if="$(arg bag)" >
    
    <!-- Use simulation time for TFs -->
    <param name="use_sim_time" value="true" />

    <!-- Bag file -->
    <node name="bag_player" type="play" pkg="rosbag" args="$(arg bag_file) --clock -s $(arg bag_start)" />
    
    <!-- Uncompress images and depth -->
    <node name="republish_img" type="republish" pkg="image_transport" output="screen" args="compressed in:=/$(arg camera)/rgb/image_raw raw out:=/$(arg camera)/rgb/image_raw"/>
    <node name="republish_depth" type="republish" pkg="image_transport" output="screen" args="compressedDepth in:=/$(arg camera)/depth_registered/image_raw raw out:=/$(arg camera)/depth_registered/image_raw"/>
     
  </group>
  
  <!-- Launch RGBD odometry -->
  <node name="rgbd_odom" type="rgbd_odom_node" pkg="rgbd_odom" >
    <param name="camera_topic" value="/$(arg camera)" />
    <param name="tf_topic_out" value="/rgbd_odom/transform" />
    <param name="max_features" value="800" />
    <param name="flow_threshold" value="10" />
    <param name="min_matches" value="15" />
    <param name="odom_frame_id" value="map" />
    <param name="base_frame_id" value="base_link" />
    <param name="publish_image" value="true" />
    <param name="publish_point_cloud" value="true" />
    <param name="use_sba" value="false" />
    <param name="type_sba" value="1" />         # 0: MOTIONSTRUCTURE, 1: MOTION, 2: STRUCTURE (cvsba)
    <param name="max_keyframes" value="10" />
    <param name="use_imu" value="true" />
    <param name="imu_topic" value="/arduimu_v3/imu" />
    <param name="apply_downsampling" value="false" />
    <param name="pc_downsampling" value="0.1" />
    <param name="apply_crop" value="true" />
    <param name="pc_crop_dist" value="7.0" />
    <param name="apply_outlier" value="false" />
    <param name="pc_outlier_radius" value="0.4" />
    <param name="pc_outlier_neighbors" value="10" />
    <param name="ground_height_topic" value="/ground_estimator_node/ground_height" />
    <param name="ground_alpha" value="0.8" />
  </node>
  
   <!-- Launch local mapper -->
  <node name="local_mapper_node" type="local_mapper_node" pkg="mapper" >
    <remap from="cloud" to="/rgbd_odom/point_cloud" />
    <param name="map_frame_id" value="map" />
    <param name="base_frame_id" value="base_link" />
    <param name="cam_max_range" value="4.0" />
    <param name="max_poses" value="11" />
    <param name="min_translation" value="0.25" />
  </node>
  
</launch>
